text/0000-export.md

```diff
@@ -0,0 +1,401 @@
+The motivation is simply to reduce the binary size of the overall package containing all these tools.
+
+The second situation has far more use cases and currently not supported well by Rust.
+A common use case is a library that is shipped as part of the system (e.g. `libz.so` or `kernel32.dll`),
+in which case you want to use the version provided by the system the program is run on,
+and not from the system it was compiled on.
+In these cases, dynamically linking is important to make sure the library can be independently updated.
+(And it also helps to not blow up binary sizes.)
+
+We need a good solution for this second category of use cases.
+
+### Solution today
+
+Currently, a way to implement this would make use of a combination of `extern "C"`, `#[no_mangle]` and `unsafe`,
+each of which has major downsides.
+
+It'd look something like this:
+
+```rust
+//! library crate
+
+pub fn hello() {
+    println!("Hello!");
+}
+
+#[no_mangle]
+pub extern "C" fn some_unique_name_for_hello() {
+    hello();
+}
+```
+
+```rust
+//! library bindings crate
+
+#[link(name = "library")]
+extern "C" {
+    fn some_unique_name_for_hello();
+}
+
+#[inline]
+pub fn hello() {
+    unsafe { some_unique_name_for_hello() };
+}
+```
+
+```rust
+//! application crate
+
+fn main() {
+    library_bindings::hello();
+}
+```
+
+This is bad. It's very verbose and error prone. More specifically:
+
+- `#[no_mangle]` is needed to export a symbol under a stable name, but it requires manually picking a good unique name that won't collide with other items from other crates.
+- A stable ABI is necessary to allow linking code from a different compiler (version),
+  but `extern "C"` puts severe limitations on the function signatures,
+  as most Rust types can't directly pass through the C ABI.
+- `unsafe` code is required, because the compiler cannot validate the imported symbol matches the expected function signature.
+  Importing the wrong library (with the same symbol name) could result in unsoundness.
+- There are now two library crates: one that will be compiled into the dynamic library (the .dll/.so/.dylib file),
+  and one that provides the bindings to that dynamic library.
+  The second library likely fully inlined into the final application,
+  as it only has wrappers, just to bring back the original (safe) function signatures.
+
+Much of this solution could be automated by a procedural macro,
+but splitting a library crate in two falls outside of what a procedural macro can reasonably do.
+
+### Proposed solution sketch
+
+Instead of all the manual usage of `#[no_mangle]`, `extern`, and `unsafe`,
+a much better solution would look as closely as possible to the original code.
+
+With the proposal below, one only needs to add an `#[export]` attribute, and give the function a stable ABI
+(e.g. `extern "C"` or (in the future) `extern "crabi"`):
+
+```rust
+//! library crate
+
+#[export]
+pub extern "C" fn hello() {
+    println!("Hello!");
+}
```

Comment:
Would going from
```rust
#[no_mangle]
pub extern "C" fn some_unique_name_for_hello() {}
```
to
```rust
#[export]
pub extern "C" fn hello() {}
```
always be a breaking change, or would there be a way to avoid it?

In other words, is it possible to make things "work out" between the `some_unique_name_for_hello` name and the mangling done by `#[export]`?

Reply:
I think we should allow exporting the same item under multiple symbols. Maybe something like `#[export(by_path, by_path_and_hash, symbol = "a", symbol = "b")]` or something like that. I'd like to leave that as a future possibility in this RFC to keep it focussed, though. (But definitely something to consider before stabilizing `#[export]`.)